ArcTouchChallenge/ArcTouchChallenge/ViewModel/Home/HomeViewModel.swift

```diff
@@ -0,0 +1,215 @@
+                    self.delegate?.showError(message: statusMessage)
+                    return
+                }
+                if let results = object.results {
+                    self.upcomingMoviesList.append(contentsOf: results)
+                }
+            }
+            
+            self.delegate?.reloadData()
+            self.isDataLoading = false
+        }
+    }
+    
+    func getTopRated() {
+        isDataLoading = true
+        
+        let parameters = ["page": currentTopRatedListPage]
+        HomeServiceModel.shared.getTopRated(urlParameters: parameters) { (object) in
+            if let object = object as? TopRatedList {
+                if let statusMessage = object.statusMessage, statusMessage != "" {
+                    self.delegate?.showError(message: statusMessage)
+                    return
+                }
+            
+                if let results = object.results {
+                    self.topRatedList.append(contentsOf: results)
+                }
+            }
+            self.delegate?.reloadData()
+            self.isDataLoading = false
+        }
+    }
+    
+    func getPopularList() {
+        isDataLoading = true
+        
+        let parameters = ["page": currentPopularListPage]
+        HomeServiceModel.shared.getPopularList(urlParameters: parameters) { (object) in
+            if let object = object as? PopularList {
+                if let statusMessage = object.statusMessage, statusMessage != "" {
+                    self.delegate?.showError(message: statusMessage)
+                    return
+                }
+                
+                if let results = object.results {
+                    self.popularList.append(contentsOf: results)
+                }
+            }
+            self.delegate?.reloadData()
+            self.isDataLoading = false
+        }
+    }
+    
+    // MARK: - Genre methods -
+    
+    func genreTitle(at section: Int) -> String {
+        if let genre = Genre.genre(at: section) { return "  \(genre.rawValue)" }
+        return ""
+    }
+    
+    // MARK: - Movie methods -
+    
+    func numberOfMovies(at section: Int) -> Int {
+        if let genre = Genre.genre(at: section) {
+            switch genre {
+                case .upcoming: return numberOfUpcomingMovies
+                case .topRated: return numberOfTopRatedList
+                case .popular: return numberOfPopularList
+            }
+        }
+        return 0
+    }
+    
+    func isMoviesEmpty(at indexPath: IndexPath) -> Bool {
+        return numberOfMovies(at: indexPath.section) == 0
+    }
+    
+    func imageData(at section: Int, row: Int, handlerData: @escaping HandlerObject) {
+        if let results = getMoviesList(at: section) {
+            let movie = results[row]
+            
+            if let data = movie.imageData {
+                handlerData(data)
+                return
+            }
+            
+            handlerData(nil)
```

Comment:
You are calling `handlerData()` twice, is it expected?

Reply:
It was used to force loading movie image correctly stopping the reuse of cell, because, when there was a bad internet connection, it was reusing the same image for other objects.